pydantic_ai_slim/pydantic_ai/models/bedrock.py

```diff
@@ -478,17 +482,58 @@ def _map_inference_config(
 
         return inference_config
 
-    def _map_tool_config(self, model_request_parameters: ModelRequestParameters) -> ToolConfigurationTypeDef | None:
-        tools = self._get_tools(model_request_parameters)
-        if not tools:
+    def _map_tool_config(
+        self,
+        model_request_parameters: ModelRequestParameters,
+        model_settings: BedrockModelSettings | None,
+    ) -> ToolConfigurationTypeDef | None:
+        resolved = _resolve_tool_choice(model_settings, model_request_parameters)
+        function_tools = model_request_parameters.function_tools
+        output_tools = model_request_parameters.output_tools
+
+        if resolved is None:
+            tool_defs_to_send = [*function_tools, *output_tools]
+        else:
+            tool_defs_to_send = resolved.filter_tools(function_tools, output_tools)
+
+        if not tool_defs_to_send:
             return None
 
+        tools = [self._map_tool_definition(t) for t in tool_defs_to_send]
         tool_choice: ToolChoiceTypeDef
-        if not model_request_parameters.allow_text_output:
+
+        if resolved is None:
+            # Default behavior: infer from allow_text_output
+            if not model_request_parameters.allow_text_output:
+                tool_choice = {'any': {}}
+            else:
+                tool_choice = {'auto': {}}
+
+        elif resolved.mode == 'auto':
+            if not model_request_parameters.allow_text_output:
+                tool_choice = {'any': {}}
+            else:
+                tool_choice = {'auto': {}}
+
+        elif resolved.mode == 'required':
             tool_choice = {'any': {}}
-        else:
+
+        elif resolved.mode == 'none':
+            # We've already filtered to only output tools, use 'auto' to let model choose
             tool_choice = {'auto': {}}
 
+        elif resolved.mode == 'specific':
+            if not resolved.tool_names:  # pragma: no cover
+                raise RuntimeError('Internal error: resolved.tool_names is empty for specific tool choice.')
+            if len(resolved.tool_names) == 1:
+                tool_choice = {'tool': {'name': resolved.tool_names[0]}}
+            else:
+                warnings.warn("Bedrock only supports forcing a single tool. Falling back to 'any'.")
```

Comment:
Didn't we implement tool filtering exactly so we don't have to do this anymore? 😄